src/devices/wifi/nm-device-olpc-mesh.c

```diff
@@ -82,7 +82,19 @@ G_DEFINE_TYPE (NMDeviceOlpcMesh, nm_device_olpc_mesh, NM_TYPE_DEVICE)
 static gboolean
 get_autoconnect_allowed (NMDevice *device)
 {
-	return FALSE;
+	NMDeviceOlpcMesh *self = NM_DEVICE_OLPC_MESH (device);
+	NMDeviceOlpcMeshPrivate *priv = NM_DEVICE_OLPC_MESH_GET_PRIVATE (self);
+
+	/* We shall always have a companion if we're >= DISCONENCTED, and this
+	 * ought not be called until then. */
+	g_return_val_if_fail (priv->companion, FALSE);
+
+	/* We must not attempt to autoconnect when the companion is connected or
+	 * connecting, * because we'd tear down its connection. */
+	if (nm_device_get_state (priv->companion) > NM_DEVICE_STATE_DISCONNECTED)
+		return FALSE;
```

Comment:
`NMPolicy` is the one that calls `nm_device_autoconnect_allowed()`. From the patch it's not clear, that we issue an autoconnect check at the right time (for example, when the companion disconnects). In essence, whenever the return value of `get_autoconnect_allowed()` might change, we must also trigger a new check for autoconnect.

It's also not clear, that when the system starts and the OLPC device and the companion are autoconnect devices, who wins to activate first.

Shouldn't the Wi-Fi device in turn also block autoconnect when the OLCP is active?

Reply:
> `NMPolicy` is the one that calls `nm_device_autoconnect_allowed()`. From the patch it's not clear, that we issue an autoconnect check at the right time (for example, when the companion disconnects). In essence, whenever the return value of `get_autoconnect_allowed()` might change, we must also trigger a new check for autoconnect.

Makes sense. I'll look into that.

> It's also not clear, that when the system starts and the OLPC device and the companion are autoconnect devices, who wins to activate first.

That's undefined, and should be. It's about the same as if you have a hotspot and a client Wi-Fi connection at the same time. One of them wins, not clear which one.

In practice, the client connection always loses because it takes time for the scan to finish and the connection to become available.

There's one difference: the priority is not considered because it can't work across devices. Hence my remark in the commit message about the OLPC interface not being a separate device.

This is probably not a problem in practice, since when one has an autoconnect mesh connection there's use sense in also having other Wi-Fi connections autoconnectable on the same radio.

> Shouldn't the Wi-Fi device in turn also block autoconnect when the OLCP is active?

Depends. I think it sort of makes sense for a regular Wi-Fi interface to be able to preempt the mesh one (so that it acts as a fallback of sorts).

Reply:
> > Shouldn't the Wi-Fi device in turn also block autoconnect when the OLCP is active?
> 
> Depends. I think it sort of makes sense for a regular Wi-Fi interface to be able to preempt the mesh one (so that it acts as a fallback of sorts).

I'm actually wrong here. It does block the autoconnect by the means of a NM_DEVICE_AUTOCONNECT_ALLOWED signal handler, companion_autoconnect_allowed_cb(), that returns a value. (For that matter, I find signal handlers returning a value somehow backwards, but hey...)